![alt text](<../images/just enough.png>)
# Just Enough Python for AI/Data Science
## Module 4: Pandas- From Raw Data to Elegant Tables
>Pandas is every data scientist's best friend for wrangling rows and columns of messy real-world data.
### Day 11 - Summarizing and Grouping Data
----

##### Overview:

After cleaning raw data in Day 10, it's now time to **derive insights** by summarizing rows and columns and performing grouped analysis.

Today we will know how to:

- Summarize your dataset using techniques like `.describe()` and `.value_counts()`.
- Group rows by specific categories using `.groupby()` and extract meaningful insights.
- Perform aggregations, transformations, and filtering on grouped data.

#### 1. Summarizing Data

**Dataset Overview**

- Revisit the Titanic dataset



In [1]:
import pandas as pd

In [2]:
# Load Titanic dataset
df = pd.read_csv("../data/titanic_dataset.csv")  # Replace with the path to your dataset
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [3]:
# Summary overview of the data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


**Summarize Numerical Data**
- Pandas' `.describe()` method generates basic statistics for all numerical columns:

In [4]:
# Descriptive statistics for numerical columns
df.describe()


PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

You’ll see metrics like `mean`, `std` (standard deviation), `min`, `max`, and percentiles for numerical columns.

- Find why the 25% for Age is 20.125

#### 2. Grouping Data

Sometimes you need to calculate aggregated metrics **grouped by certain categories**. Use the `.groupby()` method to group data by column values.

**Basic Grouping**
- Let’s explore how many passengers survived based on their gender:

In [5]:
# Group by gender and calculate survival rate
grouped_by_gender = df.groupby("Sex")["Survived"].mean()
print("Survival Rate by Gender:\n", grouped_by_gender)

Survival Rate by Gender:
 Sex
female    0.742038
male      0.188908
Name: Survived, dtype: float64


*The result gives **average survival rates** for males and females.*

**Multiple Grouping Levels**
- Group by multiple columns, e.g., by both `Sex` and `Pclass`:

In [6]:
# Survival rates grouped by gender and passenger class
grouped_gender_class = df.groupby(["Sex", "Pclass"])["Survived"].mean()
print("\nSurvival Rate by Gender and Passenger Class:\n", grouped_gender_class)



Survival Rate by Gender and Passenger Class:
 Sex     Pclass
female  1         0.968085
        2         0.921053
        3         0.500000
male    1         0.368852
        2         0.157407
        3         0.135447
Name: Survived, dtype: float64


*This outputs a hierarchical index (multi-indexed DataFrame) showing survival rates for each combination of gender and passenger class.*

**Aggregation and Custom Functions**

- Use `.agg()` to calculate multiple metrics for grouped data:

In [7]:
# Average age and fare grouped by Pclass
grouped_aggregated = df.groupby("Pclass")[["Age", "Fare"]].agg(["mean", "max", "min"])
print("\nAggregated Statistics by Passenger Class:\n", grouped_aggregated)


Aggregated Statistics by Passenger Class:
               Age                   Fare               
             mean   max   min       mean       max  min
Pclass                                                 
1       38.233441  80.0  0.92  84.154687  512.3292  0.0
2       29.877630  70.0  0.67  20.662183   73.5000  0.0
3       25.140620  74.0  0.42  13.675550   69.5500  0.0


Here are some common aggregation tasks often applied to grouped data:

- Count: Count the number of rows in each group.
- Mean: Compute the average.
- Sum: Calculate the total sum for each group.
- Custom aggregations: Create custom logic for aggregations.

In [8]:
# Count of passengers in each class
class_count = df.groupby("Pclass")["PassengerId"].count()
print("\nPassenger Count by Class:\n", class_count)


Passenger Count by Class:
 Pclass
1    216
2    184
3    491
Name: PassengerId, dtype: int64


In [9]:
# Total and average fare grouped by embarkation point
fare_stats = df.groupby("Embarked")["Fare"].agg(["sum", "mean"])
print("\nTotal and Average Fare by Embarkation Point:\n", fare_stats)


Total and Average Fare by Embarkation Point:
                  sum       mean
Embarked                       
C         10072.2962  59.954144
Q          1022.2543  13.276030
S         17439.3988  27.079812


**Embarked** = Port of Embarkation 

C = Cherbourg

Q = Queenstown

S = Southampton

#### 3.  Filtering Grouped Data
You can filter grouped results by applying conditions.

- Example: Find groups where the average age is greater than 30:

In [10]:
grouped_age = df.groupby("Pclass").filter(lambda x: x["Age"].mean() > 30)
print("\nPassengers from Classes with Average Age > 30:\n") 
grouped_age.head()



Passengers from Classes with Average Age > 30:



PassengerId  Survived  Pclass  \
1             2         1       1   
3             4         1       1   
6             7         0       1   
11           12         1       1   
23           24         1       1   

                                                 Name     Sex   Age  SibSp  \
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
11                           Bonnell, Miss. Elizabeth  female  58.0      0   
23                       Sloper, Mr. William Thompson    male  28.0      0   

    Parch    Ticket     Fare Cabin Embarked  
1       0  PC 17599  71.2833   C85        C  
3       0    113803  53.1000  C123        S  
6       0     17463  51.8625   E46        S  
11      0    113783  26.5500  C103        S  
23      0    113788  35.5000    A6        S

In [11]:
grouped_age.shape

(216, 12)

In [12]:
grouped_age.tail()

PassengerId  Survived  Pclass  \
871          872         1       1   
872          873         0       1   
879          880         1       1   
887          888         1       1   
889          890         1       1   

                                                 Name     Sex   Age  SibSp  \
871  Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
872                          Carlsson, Mr. Frans Olof    male  33.0      0   
879     Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
887                      Graham, Miss. Margaret Edith  female  19.0      0   
889                             Behr, Mr. Karl Howell    male  26.0      0   

     Parch  Ticket     Fare        Cabin Embarked  
871      1   11751  52.5542          D35        S  
872      0     695   5.0000  B51 B53 B55        S  
879      1   11767  83.1583          C50        C  
887      0  112053  30.0000          B42        S  
889      0  111369  30.0000         C148        C

In [13]:
grouped_age.Pclass.value_counts()

Pclass
1    216
Name: count, dtype: int64

#### 4. Pivot Tables
Pivot tables are incredibly useful for summarizing data by reshaping and aggregating it.

- Example: Pivot table of survival rates by gender and passenger class:

In [14]:
# Pivot table equivalent to groupby
pivot = df.pivot_table(values="Survived", index="Sex", columns="Pclass", aggfunc="mean")
print("\nSurvival Rates (Pivot Table):\n", pivot)



Survival Rates (Pivot Table):
 Pclass         1         2         3
Sex                                 
female  0.968085  0.921053  0.500000
male    0.368852  0.157407  0.135447


----
#### Quick Exercises
1. Explore the Titanic Dataset:

    - Use .describe() to summarize the numerical columns.
    - Calculate the percentage of passengers that survived and those that didn’t.

2. Group By Analysis:

    - Group passengers by `Embarked` and calculate the total `Far`e collected from each embarkation point.
    - Group passengers by `Sex` and `Pclass`, and calculate the survival rate for each group.

3. Filter Grouped Data:

    - Find all groups (by `Pclass`) where the average passenger fare exceeds 50.
    - Filter and print all passengers who paid more than $100 and were older than 30.


**Please Note:** The solutions to above questions will be present at the end of next session's (Day 12 - Matplotlib) Notebook.


---- 


### Day 10 Exercise Solution

1. Explore the Titanic Dataset:

    - Download the Titanic dataset here https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv.
    - Load it into a Pandas DataFrame.

In [15]:
# Load Titanic dataset
df = pd.read_csv("../data/titanic_dataset.csv")  # Replace with the path to your dataset
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

2. Clean the Titanic Dataset:

    - Drop rows where Age or Embarked are missing.
    - Fill in missing Fare values with the average fare.
    - Standardize column names (e.g., make them lowercase and replace spaces with underscores).
    - Add a new column named Family_Size which is the sum of SibSp and Parch.

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [17]:
# 1. Drop rows where Age or Embarked are missing
df = df.dropna(subset=["Age", "Embarked"])

# 2. Fill in missing Fare values with the average fare
df["Fare"] = df["Fare"].fillna(df["Fare"].mean())

# 3. Standardize column names (lowercase + replace spaces with underscores)
df.columns = df.columns.str.lower().str.replace(" ", "_")

# 4. Add a new column: Family_Size = SibSp + Parch
df["family_size"] = df["sibsp"] + df["parch"]

df.head()

passengerid  survived  pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                name     sex   age  sibsp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   parch            ticket     fare cabin embarked  family_size  
0      0         A/5 21171   7.2500   NaN        S            1  
1      0          PC 17599  71.2833   C85        C            1  
2      0  STON/O2. 3101282   7.9250   NaN        S            0  
3      0            113803  53.1000  C123        S            1  
4      0            373450   8.0500   NaN        S            0

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 712 entries, 0 to 890
Data columns (total 13 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   passengerid  712 non-null    int64  
 1   survived     712 non-null    int64  
 2   pclass       712 non-null    int64  
 3   name         712 non-null    object 
 4   sex          712 non-null    object 
 5   age          712 non-null    float64
 6   sibsp        712 non-null    int64  
 7   parch        712 non-null    int64  
 8   ticket       712 non-null    object 
 9   fare         712 non-null    float64
 10  cabin        183 non-null    object 
 11  embarked     712 non-null    object 
 12  family_size  712 non-null    int64  
dtypes: float64(2), int64(6), object(5)
memory usage: 77.9+ KB


3. Sort and Filter:

    - Sort passengers by Fare in descending order.
    - Filter and print all passengers who paid more than $100 and were older than 30.


In [19]:
# 1. Sort passengers by Fare in descending order
sorted_df = df.sort_values(by="fare", ascending=False)
print("Top 5 Passengers by Fare:")
sorted_df.head()



Top 5 Passengers by Fare:


passengerid  survived  pclass                                name  \
679          680         1       1  Cardeza, Mr. Thomas Drake Martinez   
258          259         1       1                    Ward, Miss. Anna   
737          738         1       1              Lesurer, Mr. Gustave J   
438          439         0       1                   Fortune, Mr. Mark   
341          342         1       1      Fortune, Miss. Alice Elizabeth   

        sex   age  sibsp  parch    ticket      fare        cabin embarked  \
679    male  36.0      0      1  PC 17755  512.3292  B51 B53 B55        C   
258  female  35.0      0      0  PC 17755  512.3292          NaN        C   
737    male  35.0      0      0  PC 17755  512.3292         B101        C   
438    male  64.0      1      4     19950  263.0000  C23 C25 C27        S   
341  female  24.0      3      2     19950  263.0000  C23 C25 C27        S   

     family_size  
679            1  
258            0  
737            0  
438            5  
341            5

In [20]:
# 2. Filter: Passengers who paid > $100 and are older than 30
filtered_df = df[(df["fare"] > 100) & (df["age"] > 30)]
print("Passengers who paid more than $100 and were older than 30:")
filtered_df.head()

Passengers who paid more than $100 and were older than 30:


passengerid  survived  pclass  \
195          196         1       1   
215          216         1       1   
258          259         1       1   
268          269         1       1   
269          270         1       1   

                                              name     sex   age  sibsp  \
195                           Lurette, Miss. Elise  female  58.0      0   
215                        Newell, Miss. Madeleine  female  31.0      1   
258                               Ward, Miss. Anna  female  35.0      0   
268  Graham, Mrs. William Thompson (Edith Junkins)  female  58.0      0   
269                         Bissette, Miss. Amelia  female  35.0      0   

     parch    ticket      fare cabin embarked  family_size  
195      0  PC 17569  146.5208   B80        C            0  
215      0     35273  113.2750   D36        C            1  
258      0  PC 17755  512.3292   NaN        C            0  
268      1  PC 17582  153.4625  C125        S            1  
269      0  PC 17760  135.6333   C99        S            0

4. Identify Duplicates:

    - Check if the dataset has duplicates and remove them, if any.

In [21]:
# 1. Check for duplicates
duplicate_rows = df[df.duplicated()]
print(f"Number of duplicate rows: {len(duplicate_rows)}")




Number of duplicate rows: 0


# HAPPY LEARNING